File: backend/app/domain/scoring/score_calculator.py
"""Server-side V3 score formula — mirrors the WASM canonical definition.

F-ARCH-3 / B-ARCH-6: the canonical implementation lives in
``wasm/math_engine.c::compute_total_score``. When the application layer
injects ``total_score_fn`` (loaded by ``infrastructure.wasm_runtime``) we
delegate the entire computation to that single C function — both client and
server then route through the same source of truth and the duplication the
audit flagged is removed.

When ``total_score_fn`` is None (older binary without the export, or
wasmtime unavailable in the deployment) we fall back to the Python mirror
below. Parity between the WASM, the JS fallback, and this Python fallback
is enforced by ``shared/score_parity_fixtures.json`` (consumed by both
``test_score_calculator_parity.py`` and ``score-calculator.parity.test.ts``).
A change to the formula must update wasm/math_engine.c FIRST, then mirror
the algebra in this file and the JS fallback in WasmBridge.ts, then
regenerate the fixtures.

Used to verify frontend-submitted scores for anti-cheat logging. Returns
None when any required input is absent so callers can skip gracefully.

Design notes:
  V3: the score base is kill_value (volume), softened by the
  survival/first-answer exponent and scaled by the rate-blend k:
      core = kill_value**(1/sqrt(denom)) * k
  The old V2 used base=k, which ignored volume and inverted the HP penalty
  when k<1. ``recompute_total_score`` returns this canonical 7-input *core*;
  the caller multiplies by SCORE_SCALE_K and difficulty_multiplier(star) to
  get the stored/displayed total_score (see session_service._verify_score).
  kill_value=0  → core is always 0 (0**x = 0). Zero-kill runs score nothing by design.
  cost_total=0  → s2=0, alpha=1, k=s1 (no penalty; the dominant rate carries
                  the blend). The pre-Q3 piecewise weight applied a 30% penalty
                  here; the continuous alpha blend removes that cliff.
"""
from __future__ import annotations
import logging
import math
from typing import Callable

logger = logging.getLogger(__name__)

# Type alias for the injected canonical scorer. Signature mirrors the C ABI
# of compute_total_score: prep durations are pre-summed by the caller so the
# function takes flat scalars rather than a variable-length list.
TotalScoreFn = Callable[[float, float, float, float, float, float, int], float]


def recompute_total_score(
    kill_value: int | None,
    time_total: float | None,
    time_exclude_prepare: list[float] | None,
    cost_total: int | None,
    health_origin: int | None,
    health_final: int | None,
    initial_answer: bool | None,
    pow_fn: Callable[[float, float], float] = math.pow,
    total_score_fn: TotalScoreFn | None = None,
) -> float | None:
    if (
        kill_value is None
        or time_total is None
        or time_exclude_prepare is None
        or cost_total is None
        or health_origin is None
        or health_final is None
        or initial_answer is None
    ):
        return None

    prep_sum = sum(time_exclude_prepare)
    if prep_sum > time_total + 0.001:
        return None

    ia = int(initial_answer)

    # Preferred path: delegate to the WASM canonical implementation. Bit-exact
    # parity with the client's totalScore is then guaranteed by both ends
    # calling the same function compiled from wasm/math_engine.c.
    if total_score_fn is not None:
        return float(total_score_fn(
            float(kill_value),
            float(time_total),
            float(prep_sum),
            float(cost_total),
            float(health_origin),
            float(health_final),
            ia,
        ))

    # Fallback: pure-Python mirror of compute_total_score for environments
    # without the WASM export. Algebra MUST stay in lock-step with the C
    # source; parity fixtures fail loudly if it drifts.
    active_time = max(0.001, time_total - prep_sum)
    s1 = kill_value / active_time
    s2 = kill_value / cost_total if cost_total > 0 else 0.0

    # Q3: continuous K blend. The old piecewise weight (0.7/0.3 vs 0.5/0.5)
    # had a discontinuity at s1 == s2 that produced visible score jumps for
    # runs that crossed it. Weighting by alpha = s1/(s1+s2) interpolates
    # smoothly; the s1+s2 == 0 zero-kill case short-circuits to k=0.
    denom_k = s1 + s2
    alpha = (s1 / denom_k) if denom_k > 0.0 else 0.0
    k = alpha * s1 + (1.0 - alpha) * s2

    exponent_denom = 1 + (2 + health_origin - health_final - ia)
    if exponent_denom < 1:
        logger.warning(
            "score_calculator: impossible HP delta (health_final=%d > health_origin=%d); "
            "clamping exponent_denom %d → 1",
            health_final,
            health_origin,
            exponent_denom,
        )
    # Q1: sqrt-softened exponent (was 1/denom). Smooths the HP-loss penalty so
    # high-difficulty plays are no longer crushed.
    exponent = 1.0 / math.sqrt(max(1, exponent_denom))
    # V3: base is kill_value (volume), softened by the exponent and
    # scaled by k. See wasm/math_engine.c::compute_total_score for the
    # rationale (old V2 used base=k, ignoring volume and inverting the HP
    # penalty when k<1). The scale K and difficulty multiplier are applied by
    # the caller, NOT here — this returns the canonical 7-input core.
    return float(pow_fn(max(0.0, kill_value), exponent) * k)


# ── V3 post-core transforms (applied by the application layer, NOT inside the
# canonical core above, so the parity fixtures keep testing the pure 7-input
# formula). Mirrored on the frontend in domain/scoring/score-calculator.ts —
# keep both in lock-step. ──

# Magnitude constant applied on top of the canonical core. The V3 core
# (killValue**exponent * k) already spans roughly tens to ~1e5 at realistic
# kill_values — the same order of magnitude as the legacy integer ``score``
# (per-level caps 5,000-100,000) — so
# COALESCE(total_score, score) mixes old and new rows without inversion and no
# extra inflation is needed (K = 1, identity). Kept as a named, front/back-
# mirrored knob for future retuning. WARNING: K > 1 risks the TOTAL_SCORE_MAX
# (1e6) clamp on high-star runs — the L5 realistic-max core is ~135k, so even
# K=10 would slam L4/L5 into the cap and flatten top-end ranking. Pure scale —
# does NOT change ranking order.
SCORE_SCALE_K = 1.0


def difficulty_multiplier(star_rating: int) -> float:
    """Reward higher-difficulty (higher star) runs. 1★→1.0 … 5★→2.0.

    0.25 and the integer (star-1) are exact in IEEE-754, so this multiplier is
    a bit-exact double on both the frontend and the backend — the server-side
    application of K * difficulty stays reproducible across languages.
    """
    return 1.0 + 0.25 * (star_rating - 1)

File: frontend/src/domain/scoring/score-calculator.ts
// F-ARCH-3 / B-ARCH-6: the canonical V3 score CORE lives in
// wasm/math_engine.c::compute_total_score. This module computes the s1/s2/k/
// exponent breakdown locally for the ScoreResultView display, but the core
// value always comes from `computeTotalScoreWasm` so it agrees with the
// server-side wasmtime-py recomputation to the last ULP. When the WASM module
// hasn't been rebuilt with the export yet, the bridge's JS fallback runs the
// same algebra and parity is enforced by shared/score_parity_fixtures.json
// (consumed by both sides).
//
// `totalScore` here is the DISPLAY value = core × SCORE_SCALE_K ×
// difficultyMultiplier(star). The server applies the identical transform from
// the trusted session row, so this equals the stored/leaderboard total_score.
// The modern client does NOT submit total_score — the server is the
// sole authority — so this scaling is display-only and never gates a session.
//
// Design notes:
//   V3: the score base is killValue (volume), softened by the
//     survival/first-answer exponent and scaled by the rate-blend k. The old
//     V2 used base=k, which ignored volume and inverted the HP penalty when
//     k<1. computeTotalScoreWasm returns this canonical 7-input *core*; this
//     module then multiplies by SCORE_SCALE_K and difficultyMultiplier(star)
//     to get the displayed totalScore, mirroring the server
//     (session_service._verify_score). This is display-only: the modern client
//     does not submit total_score, so the server stays the sole authority.
//   killValue=0  → totalScore is always 0 (0**x = 0). Zero-kill runs score nothing by design.
//   costTotal=0  → s2=0, alpha=1, k=s1 (no penalty; the dominant rate carries
//                  the blend). The pre-Q3 piecewise weight applied a 30% penalty
//                  here; the continuous alpha blend removes that cliff.
//   mUsed is a debug field only; the backend anti-cheat verifier does not track it.
import { computeTotalScoreWasm } from '@/math/WasmBridge'

// V3 post-core transforms — mirror backend domain/scoring/score_calculator.py
// (SCORE_SCALE_K + difficulty_multiplier). Keep both in lock-step: the server
// applies the identical scale so the end-of-game total matches the stored /
// leaderboard value. 0.25·(star−1) is exact in IEEE-754, so the multiplier is
// bit-identical across JS and Python.
//
// K = 1 (identity): the core already lands in the thousands-to-~100k range at
// realistic kill_values. K > 1 risks the server's TOTAL_SCORE_MAX (1e6) clamp
// on high-star runs and would flatten top-end ranking — see the backend note.
export const SCORE_SCALE_K = 1
export function difficultyMultiplier(starRating: number): number {
  return 1 + 0.25 * (starRating - 1)
}

export interface ScoreInput {
  killValue: number
  timeTotal: number               // seconds
  timeExcludePrepare: number[]    // duration of each prep phase in seconds
  costTotal: number
  healthOrigin: number
  healthFinal: number
  initialAnswer: 0 | 1
  starRating: number              // 1–5; drives the difficulty multiplier
}

export interface ScoreBreakdown {
  s1: number
  s2: number
  k: number
  exponent: number
  totalScore: number
  activeTime: number
  mUsed: number
}

export function calculateScore(input: ScoreInput): ScoreBreakdown {
  const prepTimeSum = input.timeExcludePrepare.reduce((a, b) => a + b, 0)
  const activeTime = Math.max(0.001, input.timeTotal - prepTimeSum)

  const s1 = input.killValue / activeTime
  const s2 = input.costTotal > 0 ? input.killValue / input.costTotal : 0

  // Q3: continuous K blend. The old piecewise weight (0.7/0.3 vs 0.5/0.5) had
  // a discontinuity at s1 == s2; alpha = s1/(s1+s2) interpolates smoothly.
  // The zero-kill case (s1+s2 == 0) short-circuits to k=0 to keep 0**x = 0.
  // mUsed (debug-only field) reports the s1-side weight so legacy tooltips
  // still surface "how much of the score came from time vs cost".
  const denomK = s1 + s2
  const alpha = denomK > 0 ? s1 / denomK : 0
  const k = alpha * s1 + (1 - alpha) * s2
  const mUsed = alpha

  // Invariant: healthOrigin must equal INITIAL_HP (set by createInitialState in GameState.ts).
  // Under normal game rules healthFinal ≤ healthOrigin, so rawExponentDenom ≥ 2 and the clamp
  // below never fires. The only path to rawExponentDenom < 1 is backend data that reports a
  // session-end HP above the session-start HP — impossible without a mechanic that raises maxHp
  // beyond INITIAL_HP, which does not exist.
  const rawExponentDenom = 1 + (2 + input.healthOrigin - input.healthFinal - input.initialAnswer)
  if (rawExponentDenom < 1) {
    console.warn(
      `score-calculator: impossible HP delta (healthFinal=${input.healthFinal} > healthOrigin=${input.healthOrigin}); clamping exponentDenom ${rawExponentDenom} → 1`,
    )
  }
  const exponentDenom = Math.max(1, rawExponentDenom)
  // Q1: sqrt-softened exponent (was 1/denom). Smooths the HP-loss penalty.
  const exponent = 1 / Math.sqrt(exponentDenom)
  // Source-of-truth score lives in WASM (wasm/math_engine.c). The breakdown
  // above is recomputed locally only because ScoreResultView displays
  // intermediate fields; the value sent to/verified by the server is the
  // WASM-derived totalScore.
  const core = computeTotalScoreWasm(
    input.killValue,
    input.timeTotal,
    prepTimeSum,
    input.costTotal,
    input.healthOrigin,
    input.healthFinal,
    input.initialAnswer,
  )
  // V3: lift the core into the player-facing range and apply difficulty. The
  // server applies the identical transform from the trusted star_rating, so
  // this displayed value equals the stored/leaderboard total_score.
  const totalScore = core * SCORE_SCALE_K * difficultyMultiplier(input.starRating)

  return {
    s1: round4(s1),
    s2: round4(s2),
    k: round4(k),
    exponent: round4(exponent),
    totalScore: round4(totalScore),
    activeTime: round4(activeTime),
    mUsed,
  }
}

function round4(v: number): number {
  return Math.round(v * 10000) / 10000
}
